File: scripts/scrape_quvi_coberturas.js
#!/usr/bin/env node
const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');

const OUT = path.join(__dirname, '..', 'adjuntos', 'quvi_coberturas.csv');
const FAILED = path.join(__dirname, '..', 'adjuntos', 'quvi_cobertura_failed.txt');
const ISAPRES = ['banmedica', 'colmena', 'consalud', 'cruz-blanca', 'esencial', 'nueva-masvida', 'vida-tres'];

function save(data) {
    if (!data.length) return;
    const keys = Object.keys(data[0]);
    fs.writeFileSync(OUT, [keys.join(','), ...data.map(r => keys.map(k => r[k] || '').join(','))].join('\n'));
}

(async () => {
    const browser = await chromium.launch({ headless: true });
    const ctx = await browser.newContext({ viewport: { width: 1440, height: 900 } });
    const page = await ctx.newPage();
    const results = [], failed = [];

    for (const isapre of ISAPRES) {
        console.log(`\n=== ${isapre.toUpperCase()} ===`);
        await page.goto(`https://www.quvi.cl/isapres/${isapre}`, { waitUntil: 'networkidle', timeout: 30000 });
        await page.waitForTimeout(3000);

        // Debug: dump page title and all links/buttons
        const title = await page.title();
        console.log(`  Page title: ${title}`);

        // Find plan cards — try multiple selectors
        const cards = await page.evaluate(() => {
            // Collect all plan links/cards
            const items = [];
            // Try data-plan-code
            document.querySelectorAll('[data-plan-code]').forEach(el => {
                items.push({ code: el.getAttribute('data-plan-code'), text: el.textContent?.trim().slice(0,40) });
            });
            if (items.length) return items;
            // Try links to /plan/
            document.querySelectorAll('a[href*="/plan/"]').forEach(el => {
                const href = el.getAttribute('href');
                const code = href.split('/plan/')[1]?.split('/')[0]?.split('?')[0];
                if (code && code.length > 2) items.push({ code, text: el.textContent?.trim().slice(0,40) });
            });
            return items;
        });

        console.log(`  Found ${cards.length} plan cards`);

        for (let i = 0; i < cards.length; i++) {
            const { code } = cards[i];
            try {
                // Click the card/link to navigate to plan detail
                await page.goto(`https://www.quvi.cl/plan/${code}`, { waitUntil: 'networkidle', timeout: 15000 });
                await page.waitForTimeout(1500);

                // Try to extract coverage from the plan detail page
                const data = await page.evaluate(() => {
                    // Method 1: Check for SEO data section
                    const seo = document.querySelector('.seo-plan-data');
                    // Method 2: Check for any coverage text in the full page
                    const body = document.body?.textContent || '';
                    
                    const extract = (label) => {
                        const r = new RegExp(label + '[:\s]*([^\n]{2,50})', 'i');
                        const m = body.match(r);
                        return m ? m[1].trim() : '';
                    };

                    // Method 3: Look for structured coverage elements
                    let hosp = extract('Cobertura hospitalaria');
                    let amb = extract('Cobertura ambulatoria');
                    
                    // Fallback: try extracting from badge/card elements
                    if (!hosp) {
                        const hospEl = document.querySelector('[data-criterio="hospitalaria"]');
                        if (hospEl) hosp = hospEl.textContent?.trim();
                    }

                    return {
                        nombre: document.querySelector('h1')?.textContent?.trim() || '',
                        cobertura_hospitalaria: hosp,
                        cobertura_ambulatoria: amb,
                        nota_global: extract('Nota Global'),
                        tope_anual: extract('Tope anual'),
                        prestadores: extract('prestadores'),
                        costo_uf: extract('UF'),
                        linea: extract('Línea'),
                    };
                });

                if (data.cobertura_hospitalaria || data.cobertura_ambulatoria) {
                    data.codigo = code;
                    data.isapre = isapre;
                    data.url = `https://www.quvi.cl/plan/${code}`;
                    results.push(data);
                    console.log(`  [${i+1}/${cards.length}] ${code}: H=${data.cobertura_hospitalaria} A=${data.cobertura_ambulatoria}`);
                } else {
                    failed.push(code);
                    // Debug: dump some page text
                    const sample = await page.evaluate(() => document.body?.textContent?.slice(0, 200));
                    console.log(`  [${i+1}/${cards.length}] ${code}: sin datos. Sample: ${sample?.slice(0,80)}`);
                }

            } catch (e) {
                failed.push(code);
                console.log(`  [${i+1}/${cards.length}] ${code}: error`);
            }

            if (results.length % 20 === 0) {
                save(results);
                fs.writeFileSync(FAILED, failed.join('\n'));
            }
        }
        save(results);
        fs.writeFileSync(FAILED, failed.join('\n'));
    }

    await browser.close();
    save(results);
    console.log(`\nDone. ${results.length} with data, ${failed.length} failed.`);
})();

File: scripts/scrape_quvi_coberturas.py
#!/usr/bin/env python3
"""Scrape coberturas reales de quvi.cl usando Playwright. Background task ~20-30 min."""

import csv, time, os

urls = []
with open('adjuntos/quvi_plan_urls.txt') as f:
    for line in f:
        line = line.strip()
        if line and 'quvi.cl/plan/' in line:
            parts = line.split(',')
            code = parts[0].strip() if len(parts) > 1 else ''
            url = parts[-1].strip() if ',' in line else line
            if not url.startswith('http'):
                url = f'https://www.quvi.cl/plan/{code}'
            urls.append((code, url))

print(f"Total plans: {len(urls)}")

from playwright.sync_api import sync_playwright, TimeoutError as PlaywrightTimeout

def save_results(data):
    if not data:
        return
    keys = ['codigo', 'url'] + sorted(set().union(*(d.keys() for d in data)) - {'codigo', 'url'})
    with open('adjuntos/quvi_coberturas.csv', 'w', newline='') as f:
        writer = csv.DictWriter(f, fieldnames=keys, extrasaction='ignore')
        writer.writeheader()
        writer.writerows(data)

results = []
FAILED_FILE = 'adjuntos/quvi_cobertura_failed.txt'
failed = []

with sync_playwright() as p:
    browser = p.chromium.launch(headless=True)
    context = browser.new_context(
        user_agent='Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7)',
        viewport={'width': 1280, 'height': 800}
    )
    page = context.new_page()

    for i, (code, url) in enumerate(urls):
        if not code:
            code = url.rstrip('/').split('/')[-1]
        try:
            page.goto(url, wait_until='networkidle', timeout=15000)
            time.sleep(1.0)
            coberturas = page.evaluate("""() => {
                const result = {};
                const section = document.getElementById('pdViewCobertura');
                if (!section) return result;
                const text = section.textContent;
                const patterns = {
                    'hospitalaria': /[Hh]ospitalaria[:\s]*(\d{1,3})\s*%/,
                    'ambulatoria': /[Aa]mbulatoria[:\s]*(\d{1,3})\s*%/,
                    'caec': /CAEC[:\s]*(\d{1,3})\s*%/,
                    'urgencia': /[Uu]rgencia[:\s]*(\d{1,3})\s*%/,
                    'medicamentos': /[Mm]edicamentos?[:\s]*(\d{1,3})\s*%/,
                };
                for (k, ptn of Object.entries(patterns)) {
                    const m = text.match(ptn);
                    if (m) result[k] = parseInt(m[1]);
                }
                return result;
            }""")
            if coberturas:
                results.append({'codigo': code, 'url': url, **coberturas})
                print(f"[{i+1}/{len(urls)}] {code}: {coberturas}")
            else:
                failed.append(code)
                print(f"[{i+1}/{len(urls)}] {code}: sin datos")
        except PlaywrightTimeout:
            failed.append(code); print(f"[{i+1}/{len(urls)}] {code}: timeout")
        except Exception as e:
            failed.append(code); print(f"[{i+1}/{len(urls)}] {code}: {e}")
        
        if (i+1) % 100 == 0:
            save_results(results)
            with open(FAILED_FILE, 'w') as f: f.write('\n'.join(failed))

    browser.close()

save_results(results)
with open(FAILED_FILE, 'w') as f: f.write('\n'.join(failed))
print(f"Done. {len(results)} with data, {len(failed)} failed.")
